mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs

```diff
@@ -0,0 +1,619 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace Mono.Profiler.Log {
+
+	public sealed class LogProcessor {
+
+		public LogReader Reader { get; }
+
+		public LogEventVisitor Visitor { get; }
+
+		readonly LogEventPool _pool = new LogEventPool ();
+
+		LogStreamHeader _streamHeader;
+
+		LogBufferHeader _bufferHeader;
+
+		ulong _time;
+
+		bool _done;
+
+		public LogProcessor (LogReader reader, LogEventVisitor visitor)
+		{
+			if (reader == null)
+				throw new ArgumentNullException (nameof (reader));
+
+			if (visitor == null)
+				throw new ArgumentNullException (nameof (visitor));
+
+			Reader = reader;
+			Visitor = visitor;
+		}
+
+		public void Process ()
+		{
+			Process (CancellationToken.None);
+		}
+
+		public void Process (CancellationToken token)
+		{
+			if (_done)
+				throw new InvalidOperationException ("This processor cannot be reused.");
+
+			_streamHeader = new LogStreamHeader (Reader);
+
+			var events = new List<LogEvent> ();
+
+			void ProcessEvents ()
+			{
+				foreach (var ev in events.OrderBy (x => x.Timestamp)) {
+					token.ThrowIfCancellationRequested ();
+
+					ev.Accept (Visitor);
+					_pool.Return (ev);
+				}
+
+				events.Clear ();
+			}
+
+			try {
+				while (true) {
+					token.ThrowIfCancellationRequested ();
+
+					_bufferHeader = new LogBufferHeader (_streamHeader, Reader);
+
+					var goal = Reader.BaseStream.Position + _bufferHeader.Length;
```

Comment:
I believe we must have seekable and non-seekable API versions. There is no way to have performant "rich" API built on top of non-seekable API because we put everything in memory to emulate that.